multisig/src/lib.rs

```diff
@@ -0,0 +1,348 @@
+use std::collections::HashSet;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_sdk::collections::Map;
+use near_sdk::json_types::{Base58PublicKey, Base64VecU8, U128, U64};
+use near_sdk::{env, near_bindgen, AccountId, Promise, PromiseOrValue, PublicKey};
+use serde::{Deserialize, Serialize};
+
+/// Unlimited allowance for multisig keys.
+const DEFAULT_ALLOWANCE: u128 = 0;
+
+pub type RequestId = u32;
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize)]
+pub enum MultiSigRequest {
+    Transfer {
+        receiver_id: AccountId,
+        amount: U128,
+    },
+    AddKey {
+        public_key: Base58PublicKey,
+    },
+    DeleteKey {
+        public_key: Base58PublicKey,
+    },
+    FunctionCall {
+        contract_id: AccountId,
+        method_name: String,
+        args: Base64VecU8,
+        deposit: U128,
+        gas: U64,
+    },
+    SetNumConfirmations {
+        num_confirmations: u32,
+    },
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct MultiSigContract {
+    num_confirmations: u32,
+    request_nonce: RequestId,
+    requests: Map<RequestId, MultiSigRequest>,
+    confirmations: Map<RequestId, HashSet<PublicKey>>,
+}
+
+impl Default for MultiSigContract {
+    fn default() -> Self {
+        env::panic(b"Multisig contract should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl MultiSigContract {
+    /// Initialize multisig contract.
+    /// @params num_confirmations: k of n signatures required to perform operations.
+    #[init]
+    pub fn new(num_confirmations: u32) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            num_confirmations,
+            request_nonce: 0,
+            requests: Default::default(),
+            confirmations: Default::default(),
```

Comment:
I prefer to explicitly give prefixes for the collections to avoid conflicts if later you have to remove something or change the order.

```suggestion
            requests: Map::new(b"r".to_vec()),
            confirmations: Map::new(b"c".to_vec()),
```